core/components/component.js

```diff
@@ -0,0 +1,816 @@
+
+  /**
+   * Error when an operation requiring DOM interaction is made when the
+   * component is not in the document
+   */
+  NOT_IN_DOCUMENT: 'Operation not supported while component is not in document'
+};
+
+/**
+ * Set the default right-to-left value. This causes all component's created from
+ * this point forward to have the given value. This is useful for cases where
+ * a given page is always in one directionality, avoiding unnecessary
+ * right to left determinations.
+ * @param {?boolean} rightToLeft Whether the components should be rendered
+ *     right-to-left. Null iff components should determine their directionality.
+ */
+Blockly.Component.setDefaultRightToLeft = function(rightToLeft) {
+  Blockly.Component.defaultRightToLeft_ = rightToLeft;
+};
+
+/**
+ * Gets the unique ID for the instance of this component.  If the instance
+ * doesn't already have an ID, generates one on the fly.
+ * @return {string} Unique component ID.
+ */
+Blockly.Component.prototype.getId = function() {
+  return this.id_ || (this.id_ = this.idGenerator_.getNextUniqueId());
+};
+
+/**
+ * Assigns an ID to this component instance.  It is the caller's responsibility
+ * to guarantee that the ID is unique.  If the component is a child of a parent
+ * component, then the parent component's child index is updated to reflect the
+ * new ID; this may throw an error if the parent already has a child with an ID
+ * that conflicts with the new ID.
+ * @param {string} id Unique component ID.
+ */
+Blockly.Component.prototype.setId = function(id) {
+  if (this.parent_ && this.parent_.childIndex_) {
+    // Update the parent's child index.
+    delete this.parent_.childIndex_[this.id_];
+    this.parent_.childIndex_[id] = this;
+  }
+
+  // Update the component ID.
+  this.id_ = id;
+};
+
+/**
+ * Gets the component's element.
+ * @return {Element} The element for the component.
+ */
+Blockly.Component.prototype.getElement = function() {
+  return this.element_;
+};
+
+/**
+ * Sets the component's root element to the given element.  Considered
+ * protected and final.
+ *
+ * This should generally only be called during createDom. Setting the element
+ * does not actually change which element is rendered, only the element that is
+ * associated with this UI component.
+ *
+ * This should only be used by subclasses and its associated renderers.
+ *
+ * @param {Element} element Root element for the component.
+ */
+Blockly.Component.prototype.setElementInternal = function(element) {
+  this.element_ = element;
+};
+
+/**
+ * Returns an array of all the elements in this component's DOM with the
+ * provided className.
+ * @param {string} className The name of the class to look for.
+ * @return {!IArrayLike<!Element>} The items found with the class name provided.
+ */
+Blockly.Component.prototype.getElementsByClass = function(className) {
+  return this.element_ ?
+      document.getElementsByClass(className, this.element_) :
```

Comment:
this.element_ would be null in that case, but I think this is just a nice helper method. No reason you couldn't getElement and run this elsewhere. Since we don't use it, removing..